Fixes #38916 - Update EditorHostSelect to pf5

adamlazik1 opened the pull request:
Use the shared PF5 AutocompleteInput for the template editor preview host
field.

Drop patternfly-react-extensions and its webpack vendor bundle and SCSS
imports now that the editor no longer uses the extensions Select
component.

Co-authored-by: Lukas Jezek <ljezek@redhat.com>

kfamilonidis commented on webpack/assets/javascripts/react_app/components/Editor/components/EditorHostSelect.js:
```diff
@@ -1,6 +1,18 @@
 import React, { Component } from 'react';
 import PropTypes from 'prop-types';
-import { Select } from 'patternfly-react-extensions';
+import {
```
these changes are waiting to be integrated to [this](https://github.com/theforeman/foreman/pull/10728/files#diff-7907094832ed488fcf9be5a68c3262ec7c05ad6f919c51bef1b73c54603f6d7cR38) tests.

adamlazik1 replied:
The PRs do indeed seem intertwined. Do you think this PR should be rebased on #10728 or should this PR go first, then #10728 should be rebased on top of it?

kfamilonidis replied:
It seems that #10728 should be merged first to leverage its passing tests. Subsequent changes can then be fully integrated with these tested, stable components.

kfamilonidis reviewed (commented):
Good approach. It would help to cherry-pick these changes to align with the tests already written for this part [here](https://github.com/theforeman/foreman/pull/10728/files#top).

evgeni commented on package.json:
```diff
@@ -68,7 +68,6 @@
     "multiselect": "~0.9.12",
     "number_helpers": "^0.1.1",
     "os-browserify": "^0.3.0",
-    "patternfly-react-extensions": "^3.0.15",
```
Deleting things from `package.json` will always get you a :package: ACK from me ;-)

MariaAga commented:
I didnt read the code yet, but isnt 
> replicate the behavior of https://v5-archive.patternfly.org/components/menus/select/#typeahead 

The same as this pr? https://github.com/theforeman/foreman/pull/10734

adamlazik1 commented:
> I didnt read the code yet, but isnt
> 
> > replicate the behavior of https://v5-archive.patternfly.org/components/menus/select/#typeahead
> 
> The same as this pr? #10734

@MariaAga mostly the same but EditorHostSelect has some extra properties and I don't really know whether adjusting AutocompleteInput to replace EditorHostSelect won't break the use of AutocompleteInput in other places.

pondrejk commented:
Took a brief look at the changes and also at airgun:
- I see editor-select-container stayed as an ID, which is great but I don't see it referenced in airgun anyway -- that's probably something to be changed, but this fact tells me that the tests doing template preview rely on the first host prefill, so in that matter this will definitely be a breaking change for ds tests. I'd like you to consider getting back to the fist item prefill, I think it was a nice click saver for users
- the padding issue is definitely a problem for me, imho its ok to make the navbar thicker if we cant make the dropdown thinner
- also I'd add some definition to the dropdown border, I imagine it would be hard to read, or maybe if it was with white background, not gray on gray

adamlazik1 committed:
Fixes #38916 - Update EditorHostSelect to pf5

Use the shared PF5 AutocompleteInput for the template editor preview host
field.

Drop patternfly-react-extensions and its webpack vendor bundle and SCSS
imports now that the editor no longer uses the extensions Select
component.

Co-authored-by: Lukas Jezek <ljezek@redhat.com>

adamlazik1 committed:
Update AutoCompleteInput component

Made-with: Cursor

adamlazik1 commented:
After discussion with @Lukshio we determined that it is more beneficial to first update EditorHostSelect to PF5, then update the EditorNavBar to  PF5 as the parent component.

adamlazik1 commented on webpack/assets/javascripts/react_app/components/Editor/components/EditorNavbar.js:
```diff
@@ -113,26 +122,39 @@ const EditorNavbar = ({
+                    <Spinner size="sm" loading />
+                  </span>
+                )}
+              </div>
```
This is a very rough preservation of the Spinner. @Lukshio would it make sense to implement `isLoading` behavior directly in AutoCompleteInput?

Lukshio replied:
It looks okay to me, input should be rendered once the data are loaded.

Lukshio commented on webpack/assets/javascripts/react_app/components/common/AutocompleteInput/AutocompleteInput.js:
```diff
@@ -1,6 +1,6 @@
 /* eslint-disable max-lines */
 
-import React, { useEffect, useRef, useState } from 'react';
+import React, { useEffect, useMemo, useRef, useState } from 'react';
```
Changes in this file are not necessary

adamlazik1 replied:
The search is somewhat broken without them.

### Without the changes:
- on page load:
<img width="448" height="165" alt="image" src="https://github.com/user-attachments/assets/20e55738-9a46-464d-af0b-20ba84e0c82a" />

- search "prov": 
<img width="448" height="165" alt="image" src="https://github.com/user-attachments/assets/047d78da-e0b8-4704-831c-5c408610e455" />

### With the changes:
- on page load:
<img width="415" height="260" alt="image" src="https://github.com/user-attachments/assets/e6d2a8e3-bc1b-4431-8515-04b1c422a2df" />

- search "prov":
<img width="415" height="260" alt="image" src="https://github.com/user-attachments/assets/5999586e-b46b-4386-8487-54c9a1bdebce" />

Lukshio replied:
I retested the behavior, and it looks like it's actually needed. It also prevents page from infinity rerendering.

Lukshio commented on webpack/assets/javascripts/react_app/components/Editor/components/EditorNavbar.js:
```diff
@@ -113,26 +122,39 @@ const EditorNavbar = ({
                 }
-                searchQuery={searchQuery}
-                onToggle={onHostSelectToggle}
-                onSearchChange={onHostSearch}
-                onSearchClear={onSearchClear}
-                options={isSearchingHosts ? filteredHosts : hosts}
-                key="hostsSelect"
-              />
+              >
+                <AutocompleteInput
+                  name="editor-preview-host"
```
Input is little bit overflowing, but that will be fixed in navbar PR, so that's just a nitpick

<img width="694" height="64" alt="Image" src="https://github.com/user-attachments/assets/7abe8c42-5ec7-473f-94e8-1b439f548f1f" />

adamlazik1 replied:
Yeah, I would also prefer if we dealt with this in navbar PR.

pnovotny commented:
Couple things found during PR testing:

- one scenario where the host search acts a bit strange:

If you have selected a host (given the host has ID `5`), you clear the text and type in some string that doesn't match anything, for example `ddd`.
Message `No results found for ddd` is shown (this is correct).
Now, if you click somewhere around the search input that makes the message disappear, the search string `ddd` is replaced with `5` - the host ID.
Screen capture:
[host-selector-search-id-instead-of-name.webm](https://github.com/user-attachments/assets/a46b3c69-6b3e-4790-aabd-ea6db1f53af0)

- as already commented [here](https://github.com/theforeman/foreman/pull/10777#discussion_r3062918527), 
the bottom padding of the host select and vertical alignment of `Safemode` checkbox is little bit off . But I am also fine with being it fixed in the [EditorNavbar](https://github.com/theforeman/foreman/pull/10728) PR.

Lukshio commented on webpack/assets/javascripts/react_app/components/Editor/components/EditorNavbar.js:
```diff
@@ -113,26 +122,39 @@ const EditorNavbar = ({
-                onSearchClear={onSearchClear}
-                options={isSearchingHosts ? filteredHosts : hosts}
-                key="hostsSelect"
-              />
+              >
+                <AutocompleteInput
+                  name="editor-preview-host"
+                  placeholder={__('Filter Host...')}
+                  selected={
+                    selectedHost.id === '' ? undefined : selectedHost.id
+                  }
```
```suggestion
                  selected={selectedHost.id}
```

This can be simplified.

adamlazik1 replied:
Applied.

Lukshio commented on webpack/assets/javascripts/react_app/components/Editor/components/EditorNavbar.js:
```diff
@@ -113,26 +122,39 @@ const EditorNavbar = ({
+                        renderPath: selectedRenderPath,
+                        templateKindId,
+                      });
+                    }
```
Move please all the logic inside resolveHostById function

adamlazik1 replied:
Applied.

Lukshio commented on webpack/assets/javascripts/react_app/components/Editor/components/EditorNavbar.js:
```diff
@@ -113,26 +122,39 @@ const EditorNavbar = ({
+                    selectedHost.id === '' ? undefined : selectedHost.id
+                  }
+                  options={hostOptions}
+                  onChange={onHostSearch}
```
```suggestion
```

This line is causing the issue described by @pnovotny. Old component used this as filter for input options, but autocomplete input handles this internally. So once user type some input, autocomplete gets new filtered options without the original one, so it can't pair id to label. That's why there is number onBlur.

It can be deleted with all related props, selectors and reducers.

adamlazik1 replied:
I can confirm that this fixes the behavior, but I am a little bit concerned about performance when there is a large number of hosts, let's say 10000. Won't client-side-only filtering be worse then server-side filtering in this case?

Lukshio replied:
You can simulate that behavior by creating large number of hosts manually and passing it as the options array.

adamlazik1 replied:
I ran into another issue. Turns out server only returns a batch of 10 hosts, which means that server-side filtering is here to stay. It also prevents the performance issues that I was concerned about. I am currrently looking in to other ways to fix this.

Lukshio replied:
Storing selected item as a object, not only ID should solve the issue

adamlazik1 replied:
Thanks, applied.

adamlazik1 committed:
Replace EditorHostSelect with AutocompleteInput

Use the shared PF5 AutocompleteInput for the template editor preview host
field, including the existing debounced preview_host_collection search and
filtered host list behavior.

Made-with: Cursor

adamlazik1 committed:
Remove patternfly-react-extensions

Drop the dependency and its webpack vendor bundle and SCSS imports now that
the editor no longer uses the extensions Select component.

Made-with: Cursor

adamlazik1 committed:
fixup! Update AutoCompleteInput component

adamlazik1 committed:
fixup! Replace EditorHostSelect with AutocompleteInput

Lukshio commented on webpack/assets/javascripts/react_app/components/common/AutocompleteInput/AutocompleteInput.js:
```diff
@@ -13,11 +13,18 @@ import {
   HelperTextItem,
   HelperText,
 } from '@patternfly/react-core';
+import classNames from 'classnames';
```
This is not necessary, same as using classname to hide the input.

Lukshio commented on webpack/assets/javascripts/react_app/components/common/AutocompleteInput/AutocompleteInput.js:
```diff
@@ -13,11 +13,18 @@ import {
+  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
+  label: PropTypes.string.isRequired,
+});
+
```
Why it's here and not at bottom?

Lukshio commented on webpack/assets/javascripts/react_app/components/Editor/components/EditorNavbar.js:
```diff
@@ -113,26 +155,36 @@ const EditorNavbar = ({
+                    selectedView === 'preview'
+                      ? 'editor-host-autocomplete'
+                      : 'hidden'
+                  }
```
This should be done by the showHostSelector condition, not by classname

Lukshio commented on webpack/assets/javascripts/react_app/components/common/AutocompleteInput/AutocompleteInput.js:
```diff
@@ -245,16 +289,17 @@ export const AutocompleteInputComponent = ({
   );
   return (
     <div
+      id={id}
```
Is necessary to pass id and className at the same time? also it should follow the id field pattern - `some-descritption-${name}`

Lukshio commented on webpack/assets/javascripts/react_app/components/common/AutocompleteInput/AutocompleteInput.js:
```diff
@@ -311,18 +363,23 @@ AutocompleteInputComponent.propTypes = {
 AutocompleteInputComponent.defaultProps = {
   options: [],
   selected: undefined,
+  selectedLabel: undefined,
```
What should be the pourpose? It's not used in EditorNavbar.

Lukshio reviewed (dismissed):
@adamlazik1 I think that this PR has gone too far with the changes, so now it should be better to start from the original code. 

I wrote some comments, but overall: the complexity of the code has reached unwanted state where lot of things are being rewritten/remapped.

I've created minimal patch that should solve all of the issues with minimal changes. Try to apply the patch, if it will work correctly, as it does in my setup, new testing from @pnovotny will be needed. 

[0001-minimal-patch_2.patch](https://github.com/user-attachments/files/26913215/0001-minimal-patch_2.patch)

pnovotny reviewed (approved):
> Thanks @pnovotny and @Lukshio. I squashed everything into one commit, no further changes to code were made. This is now ready for merging.

ACK

MariaAga reviewed (approved):
Did a quick look at the code and going to trust @Lukshio and @pnovotny acks.
Thanks everyone!

MariaAga commented:
@theforeman/packaging:

> Merging is blocked
> Waiting on code owner review from theforeman/packaging.

@evgeni 
[on Nov 21, 2025](https://github.com/theforeman/foreman/pull/10777#discussion_r2549538475)
Deleting things from package.json will always get you a 📦 ACK from me ;-)
